chore: Adds DependsOn converters and logic

xar-tol opened the pull request:
Adds dependency logic and conversion functions.

Addresses logic/conversions for #2175 .

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

xar-tol committed:
adds dependency validation logic

xar-tol committed:
adjusts sidecar parameters due to adjusted conversion logic

xar-tol committed:
simplifies booleans

huanjani commented on internal/pkg/deploy/cloudformation/stack/validate.go:
```diff
@@ -29,6 +30,15 @@ var (
 	errInvalidUIDGIDConfig          = errors.New("must specify both UID and GID, or neither")
 	errInvalidEFSConfig             = errors.New("bad EFS configuration: cannot specify both bool and config")
 	errReservedUID                  = errors.New("UID must not be 0")
+	errCircularDependency           = errors.New("Bad dependency name: circular dependency present")
```
nit: in golang, error strings conventionally begin with a lower case letter

xar-tol replied:
Adjusted to lowercase

huanjani commented on internal/pkg/deploy/cloudformation/stack/validate_test.go:
```diff
@@ -152,3 +152,282 @@ func Test_validateEFSConfig(t *testing.T) {
+				DependsOn: map[string]string{
+					"sidecar2": "complete",
+				},
+			},
```
what is the `wantErr` for this one?

huanjani commented on internal/pkg/deploy/cloudformation/stack/transformers.go:
```diff
@@ -77,11 +83,23 @@ func convertSidecar(s map[string]*manifest.SidecarConfig) ([]*template.SidecarOp
 }
 
+// convertDependsOn converts an Image DependsOn field to a template DependsOn version
+func convertDependsOn(i *manifest.Image, s map[string]*manifest.SidecarConfig, m string) (map[string]string, error) {
```
This function isn't being called anywhere (yet), right? It'll go in `convertSidecar()`?

xar-tol replied:
This is the conversion for the DependsOn field from the manifest.image, so it would go into the jobs/services (ex: lb_web_svc.go). Should I rename it to convertImageDependsOn?

xar-tol replied:
I currently have adding the image DependsOn saved for when I do integration tests.  Should it be added to the jobs now?

huanjani replied:
Ah, sorry, yes-- I was getting the two Dependsons confused! Probably just me, but renaming for clarity wouldn't hurt! :)

huanjani replied:
>

> I currently have adding the image DependsOn saved for when I do integration tests. Should it be added to the jobs now?

You can add it in the next PR!

xar-tol replied:
Clarified naming to convertImageDependsOn

huanjani reviewed (approved):
This looks really great! Very thorough unit tests!

huanjani commented on internal/pkg/deploy/cloudformation/stack/validate.go:
```diff
@@ -100,6 +110,135 @@ func validateSidecarMountPoints(in []manifest.SidecarMountPoint) error {
+
+func validateNoCircularDependencies(sidecars map[string]*manifest.SidecarConfig, img manifest.Image, manifestName string) error {
+	// don't let anything circularly depend on each other
+	dependencies := buildDependencyGraph(sidecars, img, manifestName)
```
This is a neat library!

xar-tol committed:
adjusts formatting

clarifies convertDependsOn
adjusts errors to be lowercase

iamhopaul123 commented on internal/pkg/deploy/cloudformation/stack/transformers.go:
```diff
@@ -51,10 +51,13 @@ var (
 
 // convertSidecar converts the manifest sidecar configuration into a format parsable by the templates pkg.
-func convertSidecar(s map[string]*manifest.SidecarConfig) ([]*template.SidecarOpts, error) {
+func convertSidecar(s map[string]*manifest.SidecarConfig, i manifest.Image, m string) ([]*template.SidecarOpts, error) {
```
Can we define a data structure for the input so that it would be more clear to the caller? For example
```go
type convertSidecar Opts {
  sidecarConfig sidecarConfig
  imageConfig manifest.Image
  workloadName string
}
```
and let's use `workloadName` instead of `manifestName`.

xar-tol replied:
Adjusted

iamhopaul123 commented on internal/pkg/deploy/cloudformation/stack/transformers.go:
```diff
@@ -77,11 +83,23 @@ func convertSidecar(s map[string]*manifest.SidecarConfig) ([]*template.SidecarOp
 }
 
+// convertDependsOn converts an Image DependsOn field to a template DependsOn version
+func convertImageDependsOn(i *manifest.Image, s map[string]*manifest.SidecarConfig, m string) (map[string]string, error) {
```
this seems to be an unused function.

xar-tol replied:
This will be used in a future PR in the jobs when I add integration tests, unless it should be added now.

iamhopaul123 replied:
Oooh could you add it when you use it? Otherwise it would be difficult to review and track.

xar-tol replied:
Added to jobs

iamhopaul123 commented on internal/pkg/deploy/cloudformation/stack/validate.go:
```diff
@@ -100,6 +110,135 @@ func validateSidecarMountPoints(in []manifest.SidecarMountPoint) error {
+
+func getNode(name string, nodes map[string]int, last int) (int, map[string]int, int) {
+	if nodes[name] != 0 {
+		return nodes[name] - 1, nodes, last
```
I would recommend using `map[string]*int` so that 0 can mean the real node index rather than if this value is set. Then we might be able to ditch this `getNode` function, which is not very easy to understand how it works.

xar-tol replied:
Changed data structure at @bvtujo 's request, no longer applies.

iamhopaul123 commented on internal/pkg/deploy/cloudformation/stack/validate.go:
```diff
@@ -29,6 +30,15 @@ var (
+
+var (
+	// Container dependency status options
+	DependsOnStatus = []string{"start", "complete", "success"}
```
Can we make "start"... constants?  Also how about "healthy"? https://github.com/aws/copilot-cli/pull/2419 should be able to unblock it.

xar-tol replied:
Is container health check in scheduled job? I believe the blocker was both the LB web service and the scheduled job.

iamhopaul123 replied:
good point. i'll adjust the PR to add schedule job as well. maybe ignore this comment until they are all fixed

xar-tol replied:
Since you added it to your PR, I added healthy.

bvtujo replied:
Does this mean that we need to add some extra validation to stop main containers from depending on `sidecar: healthy` conditions? Since sidecars don't have container health checks there's no way to determine sidecar health.

xar-tol replied:
Adjusted methods for finding essential containers and valid status to consider that sidecars can't have healthy condition.

iamhopaul123 commented on internal/pkg/deploy/cloudformation/stack/validate.go:
```diff
@@ -100,6 +110,135 @@ func validateSidecarMountPoints(in []manifest.SidecarMountPoint) error {
+			return errInvalidDependsOnStatus
+		}
+		// essential containers must have 'start' as a status
+		if name == manifestName || sidecars[name].Essential == nil || *sidecars[name].Essential {
```
```suggestion
		if name == manifestName || sidecars[name].Essential == nil || aws.BoolValue(sidecars[name].Essential) {
```
Reason is it would make it clearer that `Essential` is a `*bool` type.

bvtujo replied:
These AWS functions are also great for safe pointer dereference; `aws.BoolValue` returns false if the pointer is nil, letting you differentiate with an `== nil` check and no worry about panics.

xar-tol replied:
Adjusted

iamhopaul123 commented on internal/pkg/deploy/cloudformation/stack/validate.go:
```diff
@@ -100,6 +110,135 @@ func validateSidecarMountPoints(in []manifest.SidecarMountPoint) error {
+		// don't let a sidecar depend on itself
+		if name == sidecarName {
+			return errCircularDependency
+		}
```
circular dependencies should be checked within `validateNoCircularDependencies`. maybe we don't need this and make this function specifically for dependsOn value check?

xar-tol replied:
Removed

iamhopaul123 commented on internal/pkg/deploy/cloudformation/stack/validate.go:
```diff
@@ -100,6 +110,135 @@ func validateSidecarMountPoints(in []manifest.SidecarMountPoint) error {
+		}
+		// essential containers must have 'start' as a status
+		if name == manifestName || sidecars[name].Essential == nil || *sidecars[name].Essential {
+			if status != "start" {
```
Then it would be `start` or `healthy`?

iamhopaul123 commented on internal/pkg/deploy/cloudformation/stack/validate.go:
```diff
@@ -100,6 +110,135 @@ func validateSidecarMountPoints(in []manifest.SidecarMountPoint) error {
+	dependencies := buildDependencyGraph(sidecars, img, manifestName)
+
+	// don't let invalid containers happen
+	if dependencies == nil {
```
We should probably throw the error earlier rather than implicitly referring the `nil` return value to be this error type.

xar-tol replied:
Adjusted

iamhopaul123 reviewed (commented):
Wow this looks amazing!

xar-tol committed:
adjusts graph, addreses comments

adjusts graph to custom version
adds datastructure for convert sidecar
adds convertImageDependsOn to jobs
Adjusts essential check to use aws.bool
Removes unecessary circular dependency check
Makes invalid container errors return themselves and not nil

iamhopaul123 commented on internal/pkg/deploy/cloudformation/stack/backend_svc.go:
```diff
@@ -81,10 +81,19 @@ func (s *BackendService) Template() (string, error) {
+	convSidecarOpts := convertSidecarOpts{
+		sidecarConfig: s.manifest.Sidecars,
+		imageConfig:   &s.manifest.ImageConfig.Image,
+		workloadName:  *s.manifest.Name,
```
```suggestion
		workloadName:  aws.String(s.manifest.Name),
```

bvtujo replied:
nit: prefer `aws.*Value` functions to naked pointer dereference
```suggestion
		workloadName:  aws.StringValue(s.manifest.Name),
```

xar-tol committed:
adjusts functions for use of convertSidecarOpts

adjusts functions to use convertSidecarOpts instead of individual parameters

iamhopaul123 commented on internal/pkg/deploy/cloudformation/stack/validate.go:
```diff
@@ -29,6 +34,15 @@ var (
 	errReservedUID                  = errors.New("UID must not be 0")
+	errCircularDependency           = errors.New("bad dependency name: circular dependency present")
+	errInvalidContainer             = errors.New("container dependency does not exist")
+	errInvalidDependsOnStatus       = errors.New("container dependency status must be one of < start | complete | success >")
```
can we reuse those consts to construct this error msg?

xar-tol replied:
Adjusted.

iamhopaul123 commented on internal/pkg/deploy/cloudformation/stack/validate.go:
```diff
@@ -100,6 +114,154 @@ func validateSidecarMountPoints(in []manifest.SidecarMountPoint) error {
+	}
+
+	for name, status := range in.DependsOn {
+		// status must be one of < start | complete | success >
```
Maybe we can remove this comment as it doesn't add much value

iamhopaul123 commented on internal/pkg/deploy/cloudformation/stack/validate.go:
```diff
@@ -100,6 +114,154 @@ func validateSidecarMountPoints(in []manifest.SidecarMountPoint) error {
+	return &dependencyGraph, nil
+}
+
+type Graph struct {
```
nit this struct and its methods don't need to be public.

xar-tol replied:
Now not public.

bvtujo commented on internal/pkg/deploy/cloudformation/stack/validate.go:
```diff
@@ -29,6 +34,15 @@ var (
 	errInvalidUIDGIDConfig          = errors.New("must specify both UID and GID, or neither")
 	errInvalidEFSConfig             = errors.New("bad EFS configuration: cannot specify both bool and config")
 	errReservedUID                  = errors.New("UID must not be 0")
+	errCircularDependency           = errors.New("bad dependency name: circular dependency present")
```
can we make the circular dependency error message more informative by using `fmt.Errorf` and injecting the container names involved?

bvtujo replied:
Furthermore, can we add a separate error message for the case where a container depends on itself?

xar-tol replied:
Adjusted errors and added new message.

iamhopaul123 commented on internal/pkg/deploy/cloudformation/stack/validate.go:
```diff
@@ -100,6 +114,154 @@ func validateSidecarMountPoints(in []manifest.SidecarMountPoint) error {
+	}
+
+	for name, status := range s.imageConfig.DependsOn {
+		// status must be one of < start | complete | success >
```
same here

iamhopaul123 commented on internal/pkg/deploy/cloudformation/stack/validate.go:
```diff
@@ -100,6 +114,154 @@ func validateSidecarMountPoints(in []manifest.SidecarMountPoint) error {
+	return nil
+}
+
+func hasCycles(graph *Graph, used map[string]bool, path map[string]bool, currNode string) bool {
```
```suggestion
func (graph *Graph) hasCycles(used map[string]bool, path map[string]bool, currNode string) bool {
```
Also I would expect
```go
graph, err := buildDependencyGraph(s)
if err != nil {
  return err
}
if ok := graph.hasCycles(); !ok {
  return errCircularDependency
}
return nil
```

xar-tol replied:
Better modularized

iamhopaul123 commented on internal/pkg/deploy/cloudformation/stack/validate.go:
```diff
@@ -100,6 +114,154 @@ func validateSidecarMountPoints(in []manifest.SidecarMountPoint) error {
+	path := make(map[string]bool)
+	dependencies, err := buildDependencyGraph(s)
+
+	// don't let nonexistent containers pass
```
```suggestion
```

iamhopaul123 commented on internal/pkg/deploy/cloudformation/stack/validate.go:
```diff
@@ -100,6 +114,154 @@ func validateSidecarMountPoints(in []manifest.SidecarMountPoint) error {
+		return err
+	}
+
+	// don't let circular dependencies happen
```
```suggestion
	// don't let circular dependencies happen
```
I think one thumb of rule is if the code itself is explicit enough then we don't need additional comment (we should avoid relying too much on the comment to increase the code readability)

xar-tol replied:
Adjusted comments.

iamhopaul123 commented on internal/pkg/deploy/cloudformation/stack/validate.go:
```diff
@@ -100,6 +114,154 @@ func validateSidecarMountPoints(in []manifest.SidecarMountPoint) error {
+}
+
+func validateNoCircularDependencies(s convertSidecarOpts) error {
+	used := make(map[string]bool)
+	path := make(map[string]bool)
```
Can we define variables as close to the place where they are firstly used as possible?

bvtujo commented on internal/pkg/deploy/cloudformation/stack/validate.go:
```diff
@@ -100,6 +114,154 @@ func validateSidecarMountPoints(in []manifest.SidecarMountPoint) error {
+	path := make(map[string]bool)
+	dependencies, err := buildDependencyGraph(s)
+
+	// don't let nonexistent containers pass
```
nit: We try to use complete and grammatically correct sentences in comments wherever possible!

```suggestion
	// Don't let nonexistent containers pass.
```

xar-tol replied:
Fixed comments.

bvtujo commented on internal/pkg/deploy/cloudformation/stack/validate.go:
```diff
@@ -100,6 +114,154 @@ func validateSidecarMountPoints(in []manifest.SidecarMountPoint) error {
+	}
+
+	// don't let circular dependencies happen
+	for node := range dependencies.nodes {
```
Can you wrap this loop in a function of its own? Something like: 
```go
func (g *Graph) IsAcyclic() bool {
    used := make(map[string]bool)
    ...
    for node := range g.nodes {
        ...
    }
}
```
That way we can move the graph functions to their own file, `graph.go`, and they can be reused by other code if necessary, and a lot of code in `validateNoCircularDependencies` can be replaced by: 
```go
deps, err := buildDependencyGraph(s)
if !deps.IsAcyclic() {
    return errCircularDependency
}
```

xar-tol replied:
Better modularized.

bvtujo commented on internal/pkg/deploy/cloudformation/stack/validate.go:
```diff
@@ -100,6 +114,154 @@ func validateSidecarMountPoints(in []manifest.SidecarMountPoint) error {
+		if !isValidStatus(status) {
+			return errInvalidDependsOnStatus
+		}
+		// essential containers must have 'start' as a status
```
This will change to `start` or `healthy` as a status once we can use container health checks, right? I just want to check that I'm understanding this comment.

xar-tol replied:
Since @iamhopaul123 is adding it now, I added it to the PR

bvtujo replied:
:+1:

bvtujo commented on internal/pkg/deploy/cloudformation/stack/validate.go:
```diff
@@ -100,6 +114,154 @@ func validateSidecarMountPoints(in []manifest.SidecarMountPoint) error {
+	return nil
+}
+
+func hasCycles(graph *Graph, used map[string]bool, path map[string]bool, currNode string) bool {
```
slick recursion!

bvtujo commented on internal/pkg/deploy/cloudformation/stack/transformers_test.go:
```diff
@@ -1155,3 +1241,111 @@ func Test_convertEphemeral(t *testing.T) {
+			},
+			wantedError: errCircularDependency,
+		},
+		"invalid container dependency due to circular dependency on a sidecar": {
```
❤️

bvtujo commented on internal/pkg/deploy/cloudformation/stack/transformers_test.go:
```diff
@@ -1155,3 +1241,111 @@ func Test_convertEphemeral(t *testing.T) {
+			},
+		},
+	}
+	for name, tc := range testCases {
```
These are great unit test cases. Thank you so much for being so thorough!

bvtujo commented on internal/pkg/deploy/cloudformation/stack/transformers.go:
```diff
@@ -77,11 +83,29 @@ func convertSidecar(s map[string]*manifest.SidecarConfig) ([]*template.SidecarOp
+	return s.imageConfig.DependsOn, nil
+}
+
+type convertSidecarOpts struct {
```
Can we move this struct up to just before the definition of `convertSidecar` for readability?

xar-tol replied:
Moved.

bvtujo commented on internal/pkg/deploy/cloudformation/stack/validate.go:
```diff
@@ -100,6 +114,154 @@ func validateSidecarMountPoints(in []manifest.SidecarMountPoint) error {
+			return errInvalidDependsOnStatus
+		}
+		// essential containers must have 'start' as a status
+		if name == s.workloadName || s.sidecarConfig[name].Essential == nil || aws.BoolValue(s.sidecarConfig[name].Essential) {
```
Can we wrap this complex if statement in a small function  so that it's easier to read? that way we can write:
```go

func (s *convertSidecarOpts) isEssential(containerName string) bool {...}

if s.isEssential(name) && status != dependsOnStart {
    return errEssentialContainerStatus
}
```

This also avoids the more deeply nested conditional, which can be difficult to read and reason about.

xar-tol replied:
Better modularized

bvtujo commented on internal/pkg/deploy/cloudformation/stack/validate.go:
```diff
@@ -11,7 +11,12 @@ import (
-// Validation errors when rendering manifest into template.
+// container dependency status constants.
+const (
+	dependsOnStart    = "start"
```
What happens if someone puts "START" in their manifest, instead of "start"? Does Cloudformation accept uppercase?

iamhopaul123 replied:
I don't think so. Actually im not sure if it would be a good idea to convert all charaters to upper case

bvtujo replied:
It seems like the [CF specification](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ecs-taskdefinition-containerdependency.html#cfn-ecs-taskdefinition-containerdependency-condition) uses all uppercase values. We should convert from arbitrary capitalization in the manifest to `ALLCAPS` when we inject into the template.

xar-tol replied:
Ensured all caps on statuses

bvtujo reviewed (commented):
Alexandra. This is so good. I just have some small comments and questions about the logic

xar-tol committed:
adjusta errors, modularity, statuses

adjusts errors to state dependencies for circular dependencies, adds error for dependsny on itself
adjusts comments
adjusts status to allow healthy and adds function to ensure uppercase
better modularizes cyclic dependency detection
better modularizes essential status
 Please enter the commit message for your changes. Lines starting

xar-tol committed:
adjusts essential container status error

iamhopaul123 commented on internal/pkg/deploy/cloudformation/stack/validate.go:
```diff
@@ -100,6 +115,175 @@ func validateSidecarMountPoints(in []manifest.SidecarMountPoint) error {
+
+	// Add any idecar dependencies.
+	for name, sidecar := range s.sidecarConfig {
+		if len(sidecar.DependsOn) != 0 {
```
```suggestion
```

xar-tol committed:
removes uncessary check for dependsOn size

Lines starting

bvtujo commented on internal/pkg/deploy/cloudformation/stack/validate.go:
```diff
@@ -120,49 +121,88 @@ func validateSidecarDependsOn(in manifest.SidecarConfig, sidecarName string, s c
 
 	for name, status := range in.DependsOn {
-		// status must be one of < start | complete | success >
 		if !isValidStatus(status) {
```
Can we add a conversion to lowercase here, so that `Healthy` in the manifest can get converted here to `healthy`, register as valid, then get coverted to `HEALTHY` in the CFN template?

xar-tol replied:
All statuses are converted to uppercase before it gets into validate through the [transformations](https://github.com/aws/copilot-cli/pull/2421/files#diff-8852a201e2b451d8c315d78bf675f068153e17c99506a4dc6ecc692a25cb33cbR100), ([sidecar use](https://github.com/aws/copilot-cli/pull/2421/files#diff-8852a201e2b451d8c315d78bf675f068153e17c99506a4dc6ecc692a25cb33cbR76)) ([image use](https://github.com/aws/copilot-cli/pull/2421/files#diff-8852a201e2b451d8c315d78bf675f068153e17c99506a4dc6ecc692a25cb33cbR123)) so I don't think it's necessary, unless things can bypass that check?

bvtujo commented on internal/pkg/deploy/cloudformation/stack/validate.go:
```diff
@@ -120,49 +121,88 @@ func validateSidecarDependsOn(in manifest.SidecarConfig, sidecarName string, s c
-			return errCircularDependency
+	if !dependencies.isAcyclic() {
+		if len(dependencies.cycle) == 1 {
+			return fmt.Errorf("circular container dependency chain present: container %s depends on itself", dependencies.cycle[0])
```
Can we reword this error so it's not as deeply nested?
```suggestion
			return fmt.Errorf("container %s cannot depend on itself", dependencies.cycle[0])
```

xar-tol replied:
Adjusted.

bvtujo commented on internal/pkg/deploy/cloudformation/stack/validate.go:
```diff
@@ -171,61 +211,55 @@ func hasCycles(graph *Graph, used map[string]bool, path map[string]bool, currNod
 }
 
-func (graph *Graph) Add(fromNode, toNode string) {
+func (g *graph) Add(fromNode, toNode string) {
```
Can we make this function private too, since we aren't reusing this struct elsewhere?

xar-tol replied:
Good point! Adjusted.

bvtujo commented on internal/pkg/deploy/cloudformation/stack/validate.go:
```diff
@@ -120,49 +121,88 @@ func validateSidecarDependsOn(in manifest.SidecarConfig, sidecarName string, s c
+		if len(dependencies.cycle) == 1 {
+			return fmt.Errorf("circular container dependency chain present: container %s depends on itself", dependencies.cycle[0])
 		}
+		return fmt.Errorf("circular container dependency chain present: chain includes the following containers: %s", dependencies.cycle)
```
```suggestion
		return fmt.Errorf("circular container dependency chain includes the following containers: %s", dependencies.cycle)
```

same here, to make the error messages less deeply nested

xar-tol replied:
Adjusted.

bvtujo commented on internal/pkg/deploy/cloudformation/stack/validate.go:
```diff
@@ -120,49 +121,88 @@ func validateSidecarDependsOn(in manifest.SidecarConfig, sidecarName string, s c
+		return fmt.Errorf("circular container dependency chain present: chain includes the following containers: %s", dependencies.cycle)
 	}
 
 	return nil
```
Let's use the [return early](https://medium.com/swlh/return-early-pattern-3d18a41bba8) pattern here to make the code less nested. 

```suggestion
	if dependencies.isAcyclic() {
	    return nil
	}
	if len(dependencies.cycle) == 1 {
		return fmt.Errorf("circular container dependency chain present: container %s depends on itself", dependencies.cycle[0])
	}
	return fmt.Errorf("circular container dependency chain present: chain includes the following containers: %s", dependencies.cycle)
```

xar-tol replied:
Adjusted!

bvtujo commented on internal/pkg/deploy/cloudformation/stack/validate.go:
```diff
@@ -120,49 +121,88 @@ func validateSidecarDependsOn(in manifest.SidecarConfig, sidecarName string, s c
 }
 
-func hasCycles(graph *Graph, used map[string]bool, path map[string]bool, currNode string) bool {
+func (g *graph) isAcyclic() bool {
```
nice refactor! This is super clear.

bvtujo commented on internal/pkg/deploy/cloudformation/stack/validate.go:
```diff
@@ -171,61 +211,55 @@ func hasCycles(graph *Graph, used map[string]bool, path map[string]bool, currNod
-	if graph.nodes[fromNode] == nil {
-		graph.nodes[fromNode] = []string{}
+	// Add origin node if doesn't exist.
+	if g.nodes[fromNode] == nil {
```
Is this guaranteed to be nil if the key doesn't exist? Is that valid shorthand for:

```go
if _, ok := g.nodes[fromNode]; !ok {
```

xar-tol replied:
It appears that the comma ok format and the nil format act similarly in this case, but I adjusted it for consistency with the rest of the codebase.

xar-tol committed:
adjusts status check, errors

adjusts essential status check and valid status check to consider that sidecars can never be "HEALTHY"
adds errors for the above for sidecars
adjusts errors to be less deeply nested for circular dependencies
makes graph function private
refactors acyclic to return early

xar-tol committed:
adjusts nil format to comma ok format

iamhopaul123 reviewed (approved):
ship it :shipit: feel free to remove the label whenever the PR is ready

bvtujo commented on internal/pkg/deploy/cloudformation/stack/validate.go:
```diff
@@ -121,25 +124,33 @@ func validateSidecarDependsOn(in manifest.SidecarConfig, sidecarName string, s c
 		}
-		if isEssential(name, s) && !isEssentialStatus(status) {
-			return errEssentialContainerStatus
+		if ok, err := isEssentialStatus(status, name, s); isEssential(name, s) && !ok {
```
Can we add a quick comment here explaining this bit of logic? 
```go
// Container cannot depend on essential containers having status `complete` or `success`
```

xar-tol replied:
Added a clarifying comment.

bvtujo commented on internal/pkg/deploy/cloudformation/stack/validate.go:
```diff
@@ -121,25 +124,33 @@ func validateSidecarDependsOn(in manifest.SidecarConfig, sidecarName string, s c
-	for _, allowed := range dependsOnStatus {
-		if s == allowed {
-			return true
+func isValidStatus(status string, container string, c convertSidecarOpts) (bool, error) {
```
Can we rename this to `isValidStatusForContainer`? It might be a bit more transparent what we're checking that way.

xar-tol replied:
Adjusted naming.

bvtujo commented on internal/pkg/deploy/cloudformation/stack/validate.go:
```diff
@@ -121,25 +124,33 @@ func validateSidecarDependsOn(in manifest.SidecarConfig, sidecarName string, s c
-			return true
+func isValidStatus(status string, container string, c convertSidecarOpts) (bool, error) {
+	if _, ok := c.sidecarConfig[container]; ok {
+		for _, allowed := range sidecarDependsOnStatus {
```
can we rename this const to `sidecarDependsOnValidStatuses` or sth similar for clarity?

xar-tol replied:
Adjusted.

bvtujo commented on internal/pkg/deploy/cloudformation/stack/validate.go:
```diff
@@ -153,14 +164,20 @@ func isEssential(name string, s convertSidecarOpts) bool {
-		if s == allowed {
-			return true
+func isEssentialStatus(status string, container string, c convertSidecarOpts) (bool, error) {
+	if _, ok := c.sidecarConfig[container]; ok {
```
since this bit of code is used pretty often, we might want to refactor it into a function:

```go
func isSidecar(container string, sidecarConfig map[string]*manifest.SidecarConfig) bool
```

and this line can be replaced by:
 ```suggestion
	if isSidecar(container, c.sidecarConfig) {
```

xar-tol replied:
Better modularized.

bvtujo reviewed (approved):
Extremely nice refactor. This is such a cool little piece of code! I'm approving this because my comments are just naming nits and small readability refactors and don't reflect any major changes. Feel free to remove the do-not-merge label when you've addressed them!

xar-tol committed:
adjusts modularity of sidecar check, naming

adjusts sidecar check to be its own function
adjusts naming

xar-tol committed:
adjusts status variable naming to statuses

mergify[bot] committed:
Merge branch 'mainline' into xar-tol/sidecar-dependency-converters
